### PR TITLE
Draft canary design for discussion

### PR DESCRIPTION
**Pull request by mvs5465**
### Overview

NON-FUNCTIONAL!

This is a pseudocode-esque design example to show what the overall flow of the canary function will look like. 

They canary function will be a [CloudWatch triggered](https://docs.aws.amazon.com/lambda/latest/dg/tutorial-scheduled-events-schedule-expressions.html) Lambda function that will run once a day, sample a subset of the files that have been uploaded that day, and iterate their contents looking for invalid data.

### Functionality

Basic process:

1. List 10 files that were uploaded that day
2. Aggregate a list of all of the records in those files
3. Validate the message list
4. If there was an error, send an email alert

### Challenges and Concerns

- Configuration:
  - How do we manage the configuration of this function? How do we manage the configuration the validation library?
  - Logic will be required for handling and referencing multiple configuration files in order to handle multiple message types
  - Logic will be required for matching configuration file to record
- Sampling:
  - There is a trade off between sampling more s3 files and requiring more memory management logic
  - S3 query results can only be constrained with the prefix (filepath) and cannot be wildcarded, meaning it will require logic to query for specific filetypes
  - How do we ensure the sampling covers enough data to be confident in the results?
- Result Reporting:
  - What should be the format of the report?
  - Do we want to send reports when no validation failures occurred?
  - How can we make it easy for the user to find the invalid data?
  - What is the response process to an invalid report?
- Deployment and Integration:
  - How do we package the validation library?
  - How do we ensure the validation library is up to date?
  - Consider using Travis to automate deployment for basic CICD (credential management w/ CSN may be difficult)
  - How do we manage the automation? ([Serverless CloudFormation templates](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/what-is-sam.html) are one of the best ways)

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    if list_contains_no_errors(validation_results):
+        print("No validation errors detected.")
+    else:
+        print("Validation failed, errors detected.")
+        if SEND_EMAIL_ALERTS:
+            print("Sending email alert.")
+            send_report("Validation failed, detected %d errors out of %d messages analyzed. ")
+        else:
+            print("Sending email alert.")
+    return
```
Comment (hmusavi) [inline main.py]: Change to:
```
print("Sending` email alert.")
print_report("Validation failed, detected %d errors out of %d messages analyzed. ")
```

**Comment by hmusavi** [comment]
It's not easy to comment on the PR Description. Please place it on Confluence so we could collaborate and iterate over comments and responses.

File: package.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+mkdir package
+pip install git+https://github.com/usdot-jpo-ode/ode-output-validator-library.git --upgrade --target ./package
+zip -r function.zip main.py package/
```
Comment (hmusavi) [inline package.sh]: how about calling it `odevalidator.zip`

Comment (mvs5465) [reply package.sh]: Good call. Updated.

**Review by hmusavi — APPROVED** [review]

**Review by hmusavi — APPROVED** [review]

**Pull request merged**